algorithms-miscellaneous-3/src/main/java/com/baeldung/algorithms/breadthfirstsearch/BreadthFirstSearchAlgorithm.java

```diff
@@ -0,0 +1,61 @@
+package com.baeldung.algorithms.breadthfirstsearch;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+public class BreadthFirstSearchAlgorithm {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BreadthFirstSearchAlgorithm.class);
+
+    public static <T> Optional<Tree<T>> search(T value, Tree<T> root) {
+        Queue<Tree<T>> queue = new ArrayDeque<>();
+        queue.add(root);
+
+        return searchTreeQueue(value, queue);
+    }
+
+    private static <T> Optional<Tree<T>> searchTreeQueue(T value, Queue<Tree<T>> queue) {
+        Tree<T> currentNode;
+        while (!queue.isEmpty()) {
+            currentNode = queue.remove();
+            LOGGER.info("Visited node with value: {}", currentNode.getValue());
+
+            if (currentNode.getValue().equals(value)) {
+                return Optional.of(currentNode);
+            } else {
+                queue.addAll(currentNode.getChildren());
+            }
+        }
+
+        return Optional.empty();
+    }
```

Comment:
Why not just put all this in the search method?

Reply:
Yeah, that doesn't make much sense now that I see it again :-)